text/0000-modular-crypto-service.md

```diff
@@ -0,0 +1,164 @@
+---
+layout: default
+title: RFC Template
+nav_order: 3
+---
+
+- Feature Name: modular_crypto_service
+- Start Date: 2020-09-19
+- RFC PR: (leave this empty)
+- Fabric Component: core, fabric-sdk-*, fabric-ca
+- Fabric Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+One paragraph explanation of the feature.
+
+This RFC aims to provide broader crypto service configurability or pluggable capability. 
+It mainly changes design of crypt/X509 usage and hash algorithm opt-in in several layer.  
+
+# Motivation
+[motivation]: #motivation
+
+Why are we doing this? What use cases does it support? What is the expected
+outcome?
+
+Following is the original initiative post, translated from Chinese.
+```
+The Hyperledger Fabric code project has a very wide range of adoptions in many domestic industries and institutions, and has a very basic role in the application of the blockchain industry.
+According to Chinese policy requirements, it is necessary to support the national secret (GM) encryption algorithm. Many domestic enterprises and Fabric enthusiasts have successively carried out some corresponding national secret (GM) fabric forks with outstanding achievements. However, in this matter, there is serious duplication of working hours and waste of human resources.
```

Comment:
Can you please explain, or give a formal (English) reference to the bespoken policy? Is this policy only for intra-national networks? I'm asking because most of the world uses the (American) NIST standard - America, Australia, Europe, Africa and Asia. 

When a person from China connects to a non chinese website, with overwhelming probability, that website uses the NIST approved algorithms. So, does this policy actually take place? How does the browser of the end user know which protocol to use when connecting to a site?

Reply:
We stored a copy of one of original standard as [Financial distributed ledger technology security specification](https://github.com/Hyperledger-TWGC/fabric-gm-wiki/blob/master/%E9%87%91%E8%9E%8D%E5%88%86%E5%B8%83%E5%BC%8F%E8%B4%A6%E6%9C%AC%E6%8A%80%E6%9C%AF%E5%AE%89%E5%85%A8%E8%A7%84%E8%8C%83.pdf)
More recently, there is another related standard as [Financial application of blockchain technology -- Evaluation rules](https://github.com/Hyperledger-TWGC/fabric-gm-wiki/blob/master/%E5%8C%BA%E5%9D%97%E9%93%BE%E6%8A%80%E6%9C%AF%E9%87%91%E8%9E%8D%E5%BA%94%E7%94%A8%E4%B8%8E%E8%AF%84%E4%BC%B0%E8%A7%84%E5%88%99.pdf)

There might be a blur expression from our original initiative. Policy means some points like:
`To promote GM crypto adoption in Critical Industries`. This has been mentioned in Government for several years (from 10 years ago). 2020 we see the above specifications responding to this policy trend. 

Formal English reference does not exist theoretically. But we TWGC are also working on an Fabric Evaluation on the first specs in [Tencent doc](https://docs.qq.com/doc/DV1VMenFiQXBpeFZK). It is better formatted to apply google translate. 

About these specification applying scope, both above are limited to 'Finance', in level of 'Recommended'. So it will not apply to all intra-national network or internet. We can see it as an extension from FIPS series when coming to Financial DLT domain.

Reply:
The two links you posted are in Chinese, I can't read them. Isn't there an English version somewhere?

> About these specification applying scope, both above are limited to 'Finance', in level of 'Recommended'. So it will not apply to all intra-national network or internet. We can see it as an extension from FIPS series when coming to Financial DLT domain.

If it's recommended it means it's not mandatory, therefore since all businesses in China can still use the existing Fabric implementation, what business value does using Chinese crypto have?

Reply:
It is required in practical use case. Increasing numbers of financial institution procument will add this specification as requirement to checklist. If it is indeed valueless, we would not have heard about questions and efforts to make a fabric-GM fork. Ignoring this is not prohibited but at least harmful. 

Meanwhile I would like to highlight that this RFC does not aim to inject a native GM support in Fabric. We think for further. 

@hartm Hart, what is your opinion?

Reply:
I'm not a Fabric maintainer (or even contributor, just a free rider), so I don't think my opinion here will matter too much.  But I do like the idea of a plugin crypto architecture for Fabric, which I think would be the cleanest way to do this (Fabric maintainers, please feel free to shoot me down if I'm way off), and I'm always happy to spout off an opinion, so I'll go ahead and comment.

There are other applications that a plugin architecture would enable.  For instance, I'd like to be able to replace the core signing algorithms in Fabric with pairing-based signatures.  We could then use threshold versions of these signature schemes for a number of useful blockchain applications.  For instance, this would allow compact endorsements and small "proofs of blockchain state."  I have a bunch more applications in mind that I'd be happy to talk about if people are curious that are too long for a comment here.  These are obviously not related to government standards.

However, I think you probably want a much more detailed RFC than what is currently present (maybe I'm missing the details?).  Can you propose the exact interfaces and functions you want for modular interoperability?  I think the Fabric maintainers will potentially be much more receptive if you propose the kind of interface you want.  It looks like just 1) a generic hash primitive and 2) a generic signature primitive will be enough.  Figuring out what kinds of formats/functions will work best with Fabric will go a long way here.

Finally, doing something like this would be an ambitious project.  The more you can do it in bite sized pieces, the better.  It will make testing and security analysis much better.  So it might be best if your plan was to come up with a generic interface, put your protocols in that interface, and then slowly migrate Fabric.

Anyways, I hope I'm not ruffling any feathers by commenting here.  Thanks everyone for your hard work.

Reply:
@hartm Thanks for your valuable feedback. Indeed we are hard working to provide more details, and further changes to the RFC proposal will be pushed later.